sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
 
   PI_ASSERT(Device, PI_INVALID_DEVICE);
 
-  // Check if Device was already partitioned into the same or bigger size
-  // before. If so, we can return immediately without searching the global
-  // device cache. Note that L0 driver always returns the same handles in the
-  // same order for the given number of sub-devices.
-  if (OutDevices && NumDevices <= Device->SubDevices.size()) {
-    for (uint32_t I = 0; I < NumDevices; I++) {
-      OutDevices[I] = Device->SubDevices[I];
-      // reusing the same pi_device needs to increment the reference count
-      piDeviceRetain(OutDevices[I]);
-    }
-    if (OutNumDevices)
-      *OutNumDevices = NumDevices;
-    return PI_SUCCESS;
+  // Devices cache is normally created in piDevicesGet but still make
+  // sure that cache is populated.
+  //
+  pi_result Res = populateDeviceCacheIfNeeded(Device->Platform);
+  if (Res != PI_SUCCESS) {
+    return Res;
```

Comment:
Does this means that you always partition the devices into the maximum number of sub-devices, ignoring the user's requested number of devices?

Reply:
Yes, I have actually added assert that checks this. See line 2031.

Reply:
Oh. I see. So, initially you partition fully and keep them in the cache.
Next time if the user requests less than the maximum, we can return only the desired number of sub-devices.